src/passes/StringLowering.cpp

```diff
@@ -0,0 +1,188 @@
+//
+// StringGathering collects all string.const operations and stores them in
+// globals, avoiding them appearing in code that can run more than once (which
+// can have overhead in VMs).
+//
+// Building on that, an extended version of StringGathering will also replace
+// those new globals with imported globals of type externref, for use with the
+// string imports proposal. String operations will likewise need to be lowered.
+// TODO
+//
+
+#include <algorithm>
+
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "pass.h"
+#include "wasm-builder.h"
+#include "wasm.h"
+
+namespace wasm {
+
+struct StringGathering : public Pass {
+  // All the strings we found in the module, and a reverse mapping.
+  std::vector<Name> strings;
+  std::unordered_map<Name, Index> stringIndexes;
+
+  // Pointers to all StringConsts, so that we can replace them.
+  using StringPtrs = std::vector<Expression**>;
+  StringPtrs stringPtrs;
+
+  // Main entry point.
+  void run(Module* module) override {
+    processModule(module);
+    addGlobals(module);
+    replaceStrings(module);
+  }
+
+  // Scan the entire wasm to find the relevant strings to populate our global
+  // data structures.
+  void processModule(Module* module) {
+    struct StringWalker : public PostWalker<StringWalker> {
+      StringPtrs& stringPtrs;
+
+      StringWalker(StringPtrs& stringPtrs) : stringPtrs(stringPtrs) {}
+
+      void visitStringConst(StringConst* curr) {
+        stringPtrs.push_back(getCurrentPointer());
+      }
+    };
+
+    ModuleUtils::ParallelFunctionAnalysis<StringPtrs> analysis(
+      *module, [&](Function* func, StringPtrs& stringPtrs) {
+        if (!func->imported()) {
+          StringWalker(stringPtrs).walk(func->body);
+        }
+      });
+
+    // Also walk the global module code (for simplicity, also add it to the
+    // function map, using a "function" key of nullptr).
+    auto& globalStrings = analysis.map[nullptr];
+    StringWalker(globalStrings).walkModuleCode(module);
+
+    // Combine all the strings.
+    std::unordered_set<Name> stringSet;
+    for (auto& [_, currStringPtrs] : analysis.map) {
+      for (auto** stringPtr : currStringPtrs) {
+        stringSet.insert((*stringPtr)->cast<StringConst>()->string);
+        stringPtrs.push_back(stringPtr);
+      }
+    }
+
+    // Generate the indexes from the combined set of necessary strings, which we
+    // sort for determinism (alphabetically).
+    for (auto& string : stringSet) {
+      strings.push_back(string);
+    }
+    std::sort(strings.begin(), strings.end());
+    for (Index i = 0; i < strings.size(); i++) {
+      stringIndexes[strings[i]] = i;
+    }
+  }
+
+  // For each string index, the name of the global that replaces it.
+  std::vector<Name> globalNames;
+
+  Type nnstringref = Type(HeapType::string, NonNullable);
+
+  // Existing globals already in the form we emit can be reused. That is, if
+  // we see
+  //
+  //  (global $foo (ref null string) (string.const ..))
```

Comment:
Looks like we only reuse non-nullable globals.
```suggestion
  //  (global $foo (ref string) (string.const ..))
```